### PR TITLE
Default `jac_prototype` in ExpRK caches

### PR DESCRIPTION
**Pull request by MSeeker1340**
Fixes an issue where the Jacobian is not default initialized to a dense matrix if `has_jac(f) == true` but `f.jac_prototype == nothing` and added a regression test. Similar fixes have been applied to implicit methods before but ExpRK was left out.

### CODE REVIEW

**Comment by MSeeker1340** [comment]
The function contructed by `@ode_def` is not an ODEFunction, which makes `_compute_nl` fail. Fixed in newest commit.

File: src/perform_step/exponential_rk_perform_step.jl [hunk 1]
```diff
@@ -2,9 +2,9 @@ using LinearAlgebra: axpy!
 
 # Helper function to compute the G_nj factors for the classical ExpRK methods
 @inline _compute_nl(f::SplitFunction, u, p, t, A) = f.f2(u, p, t)
-@inline _compute_nl(f::ODEFunction, u, p, t, A) = f(u, p, t) - A * u
```
Comment (ChrisRackauckas) [inline src/perform_step/exponential_rk_perform_step.jl]: it should be outputing an `ODEFunction`?

**Comment by ChrisRackauckas** [comment]
It's an `AbstractODEFunction`. But yes, we probably don't need a restriction here

**Comment by MSeeker1340** [comment]
Rebased onto master.

**Comment by MSeeker1340** [comment]
@ChrisRackauckas Test failures seem related to floating point errors and the complex tests should not be affected by this PR. OK to merge now?

**Pull request merged**